### PR TITLE
Eliminate redundant or meaningless elaborations in type relations

### LINKED ISSUE #46900: tsc 4.5 hangs with unist-util-visit package (works fine with 4.4.4)
# Bug Report

the following example makes `tsc` hang indefinitely with version 4.5, but compiles fine with 4.4.4:

example: https://github.com/stefanprobst/issue-tsc

also see the discussion here: https://github.com/unifiedjs/unified/discussions/175 

### 🔎 Search Terms

hangs

### 🕗 Version & Regression Information

- This is a crash: `tsc` hangs, but does not give an error message
- This changed between versions 4.4.4 and  4.5.2 (same behavior also in current `@next` version)

### ⏯ Playground Link

link to a reproduction repo above

### 🙁 Actual behavior

`tsc` hangs indefinitely without error message

### 🙂 Expected behavior

`tsc` finishing compilation, or giving indication what's wrong

### LINKED ISSUE #47668: Unnecessary, misleading expansion lines for a keyof generic used in a template literal
# Bug Report

### 🔎 Search Terms

keyof generic template literal type symbol

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generics and assignability

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBA0hIHELADwBUB8UC8UAGA5sgCQDeA1vAPYBmUaAvngNwCwAUEA)

### 💻 Code

```ts
type KeyGet<T> = `get${keyof T}`;
```

### 🙁 Actual behavior

Notice the fourth `Type 'keyof T'` line onward. Seems unnecesary?

```ts
Type 'keyof T' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
  Type 'string | number | symbol' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
    Type 'symbol' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
      Type 'keyof T' is not assignable to type 'number'.
        Type 'string | number | symbol' is not assignable to type 'number'.
          Type 'string' is not assignable to type 'number'.
```

### 🙂 Expected behavior

```ts
Type 'keyof T' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
  Type 'string | number | symbol' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
    Type 'symbol' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
```

Note that if you replace `keyof T` with `string | number | symbol`, the error message is the expected behavior. [[playground link with replacement](https://www.typescriptlang.org/play?#code/C4TwDgpgBA0hIHELADwBUB8UC8UAGA5sgCQDeAzsAE4CWAdgVAD5R0CuAtgEYRXNTkQ3APYAbAL54A3AFgAUEA)]

```ts
type KeyGet<T> = `get${string | number | symbol}`;
```

```
Type 'string | number | symbol' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
  Type 'symbol' is not assignable to type 'string | number | bigint | boolean | null | undefined'.
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR removes a several pointless error elaborations, making diagnostics more concise. For example, when there is no relation between some source type and a target union type, we now only elaborate when we can identify a best match on the target side. Previously we'd default to elaborating on the last target constituent, which generally just adds noise and depends on a unstable sort order.

The PR also optimizies the `isRelatedTo` function by moving more work to the cached side on relationship checking. This improves checker performance by up to 3% and fixes an issue related to infinite constraints involving intersections.

Fixes #46900.
Fixes #47668.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test this

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by ahejlsberg** [comment]
@typescript-bot perf test this faster

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster

**Comment by RyanCavanaugh** [comment]
I haven't looked over the checker code thoroughly, but all the baseline diffs are either positive or equivalently OK IMO

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -18288,171 +18292,136 @@ namespace ts {
...
-                //    since we don't want to end up with a worse error like "`Foo` is not assignable to `NonNullable<T>`"
-                //    when dealing with generics.
-                //  * We also don't deal with primitive source types, since we already halt elaboration below.
-                if (target.flags & TypeFlags.Union && source.flags & TypeFlags.Object &&
-                    (target as UnionType).types.length <= 3 && maybeTypeOfKind(target, TypeFlags.Nullable)) {
-                    const nullStrippedTarget = extractTypesOfKind(target, ~TypeFlags.Nullable);
-                    if (!(nullStrippedTarget.flags & (TypeFlags.Union | TypeFlags.Never))) {
-                        target = getNormalizedType(nullStrippedTarget, /*writing*/ true);
+                // See if we're relating a definitely non-nullable type to a union that includes null and/or undefined
+                // plus a single non-nullable type. If so, remove null and/or undefined from the target type.
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:18296]: I actually slightly prefer the old comment, but I'm not opposed to this.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Mostly looks good. I think some comments do need to be updated, and there are some style suggestions. I am glad to see the error heuristics pick up more consistently!

**Comment by DanielRosenwasser** [comment]
> The PR also optimizies the `isRelatedTo` function by moving more work to the cached side on relationship checking.

Can you elaborate on that? I saw that you did hoist a bunch of `source.flags` and `target.flags` around in a few functions which we've heard might be a helpful optimization, but I'm not picking up what you're referring to.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -18594,6 +18549,51 @@ namespace ts {
                 return prop.valueDeclaration && container.valueDeclaration && prop.valueDeclaration.parent === container.valueDeclaration;
             }
 
+            function unionOrIntersectionRelatedTo(source: Type, target: Type, reportErrors: boolean, intersectionState: IntersectionState): Ternary {
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts:18552]: No, one or both may be a union or intersection, but it isn't known which.

**Comment by ahejlsberg** [comment]
> Can you elaborate on that?

In general, logic that sits in `isRelatedTo` is on the uncached side and logic that sits in `structuredTypeRelatedTo` is on the cached side (by way of `recursiveTypeRelatedTo`). The PR makes the following optimizations:

* Some logic is moved from `isRelatedTo` to `structuredTypeRelatedTo`, ensuring that we only execute it once for a given pair of types.
* The `isRelatedTo` function would often make *two* calls to `recursiveTypeRelatedTo` in order to check relations involving unions or intersections. That incurred the cost of computing keys and checking the relation cache twice. We now only call `recursiveTypeRelatedTo` once for a given pair of types.
* The `reportErrorResults` function is un-nested such that closures don't get allocated in `isRelatedTo`.
* The `captureErrorCalculationState` function is no longer called by `isRelatedTo` which cuts down on garbage.

We could further consider moving the excess and common property checking code to the cached side, though it isn't run all that often and would require some restructuring in `structuredTypeRelatedTo`.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -18594,6 +18549,51 @@ namespace ts {
...
+                // Source is an intersection. For the comparable relation, if the target is a primitive type we hoist the
+                // constraints of all non-primitive types in the source into a new intersection. We do this because the
+                // intersection may further constrain the constraints of the non-primitive types. For example, given a type
+                // parameter 'T extends 1 | 2', the intersection 'T & 1' should be reduced to '1' such that it doesn't
+                // appear to be comparable to '2'.
+                if (relation === comparableRelation && target.flags & TypeFlags.Primitive) {
+                    const constraints = sameMap((source as IntersectionType).types, getBaseConstraintOrType);
+                    if (constraints !== (source as IntersectionType).types) {
+                        source = getIntersectionType(constraints);
+                        if (!(source.flags & TypeFlags.Intersection)) {
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts:18576]: Actually, this is just handling the case where the intersection gets reduced down to a singleton type.

[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/compiler/checker.ts:18576]: This is the part that's handling the cases of primitives and primitive unions, but you're changing `source` for the intersection case below as well.

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

File: src/compiler/checker.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts:18576]: Yes, but the comment above the whole section already says that we're hoisting the constraints into a new intersection (with the implication that it replaces the old source). But perhaps that isn't clear enough.

**Review by DanielRosenwasser — APPROVED** [review]
Seems like a really good change! Let's get it in after the 4.6 RC gets finalized.

**Comment by mrpmorris** [comment]
@ahejlsberg 

I'm sorry to ask this here, but I have no other way to ask. Would you be willing to give your input on the proposed !! language addition to C#, and share whether or not you have any alternate ideas?

PS: I hate JS, but I really quite like using TS. I just wish it could be type-safe too.

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by RyanCavanaugh** [comment]
@typescript-bot cherry-pick this to release-4.6

**Pull request merged**